### DISCUSSION TITLE
Bem vindo ao espaço de Q&A (Perguntas e Respostas) do Curso AWS com Terraform

### DISCUSSION THREAD

**Discussion by chgasparoto** [chgasparoto/curso-aws-com-terraform#26] [Geral]
## 👋 Bem vindo(a)!
Iremos utilizar as "Discussions" como um lugar para nos conectarmos. Esperamos que você:
   * Faça perguntas sobre as quais você tenha dúvidas ou dificuldades;
   * Compartilhe ideias;
   * Seja gentil e educado(a).

Caso tenha dúvidas em como perguntar ou criar uma nova "discussion", veja o exemplo abaixo:
[Exemplo de dúvida/pergunta](https://github.com/chgasparoto/curso-aws-com-terraform/discussions/31)

**Comment by analistadeperon** [comment]
qual a configuração precisa ter para fazer o mesmo programa que vc pois nao consigo ter todas as informaçoes estaladas

**Reply by chgasparoto** [reply]
@analistadeperon eu uso o Webstorm junto com este plugin:

- https://www.jetbrains.com/webstorm/
- https://plugins.jetbrains.com/plugin/7808-hashicorp-terraform--hcl-language-support

**Reply by analistadeperon** [reply]
boa tarde nao consigo realizar as funções impossivel instalei tudo que passou mais nao abre a janela correta do projeto tem como enviar.

**Reply by chgasparoto** [reply]
Bom dia @analistadeperon, não entendi. O que você precisa exatamente?

**Comment by analistadeperon** [comment]
Bom dia um link para poder seguir as aulas pois os dados iniciais se tem github mais como colocar no sistema terraform ?

Obter o Outlook para Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by chgasparoto** [reply]
Desculpa @analistadeperon , ainda não consegui entender qual o seu problema.
O que você quer dizer com "colocar no sistema terraform"? Poderia elaborar mais, por favor?

**Comment by analistadeperon** [comment]
Não consigo acompanhar pois o sistema é diferenre preciso do curso pra trabalhar

Obter o Outlook para Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by analistadeperon** [comment]
Eu tenho a 14 funciona melhor e as empresas pedem esse padrao pra seus testes.

Obter o Outlook para Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by debroski** [comment]
Olá pessoal, estou criando na empresa onde trabalho algo simples com terraform, onde criamos de vez enquanto instancias para usar bando de dados na AWS. 
Na criação estou usando o workspace como referencia ex: "cluster-postgresql-22" este é o nome do workspace.
Não consegui identificar um formato que eu consiga pegar parte do nome do workspace, neste ex seria o numero 22.
Tem alguma dica de como posso trabalhar?

**Reply by chgasparoto** [reply]
@debroski eu criei uma pergunta para a sua dúvida, vou te responder lá https://github.com/chgasparoto/curso-aws-com-terraform/discussions/37 👍

**Comment by victorpauloti** [comment]
Ola Prof Cleber,  @chgasparoto 
Não é duvida nem pergunta vou deixar aqui uma observaçao se alguem passar por isto.
Na aula de data source  estava recebendo o seguinte erro:

`Error: Error launching source instance: InvalidParameterValue: The architecture 'x86_64' of the specified instance type does not match the architecture 'arm64' of the specified AMI. Specify an instance type and an AMI that have matching architectures, and try again. You can use 'describe-instance-types' or 'describe-images' to discover the architecture of the instance type or AMI.
        status code: 400, request id: 847c7f7b-5ce6-4dd1-9507-9af5db99f186`
   
   Estava usando a regiao us-east-1, mas parou de dar o erro e prossegui quando alterei no data.tf o name_regex para :
   name_regex = "Ubuntu" (ubuntu com a letra maiscula :). )
   Excelente curso. abs

**Reply by chgasparoto** [reply]
Fala @victorpauloti , muito obrigado pela dica!

Eu tive esse problema na criação da aplicação auto escalável. Mas agora que você falou faz todo sentido, talvez a AWS tenha mudado a primeira letra e dentro regex faz diferença.

Abraço.

**Comment by wagnerperon** [comment]
Aula [18. Built-in functions] - terraform destroy -var="env=dev"
Ola.
Fiquei com 2 dúvidas na aula  [18. Built-in functions].
Quando executo o applay para criar um ambiente seguindo as variáveis "dev" ou "prod" ele executa certo, criando o número de instâncias solicitadas.

pergunta 1:
Quando executo o applay como dev, ele cria 1 instancia - Okay
Quando executo o applay como prod, ele reaproveita a instancia já criada em dev. Não deveria "diferenciar", pois neste caso, eu perdi a instância de "dev", certo? Como faço para que isso não ocorra?

pergunta 2:
Porque quando executei terraform destroy -var="env=dev" -auto-approve ele destruiu todas as minhas instâncias? Não deveria destruir somente as de "dev", ou seja, a 1º criada?

Obrigado.

**Reply by chgasparoto** [reply]
@wagnerperon isso acontece pois você está utilizando o mesmo arquivo de `tfstate`.

Para conseguir diferenciar entre diferentes ambientes (environments) você deve usar `workspaces` ou setar um `backend`, utilizando remote state, diferente para cada ambiente na hora de iniciar o terraform.

Para o exemplo de `workspaces`, se você assistir ao vídeo você entenderá como utilizar.

Para o exemplo de diferentes `backends`, dá uma olhada nesse [repositório](https://github.com/chgasparoto/youtube-cleber-gasparoto/tree/main/0003-terraform-gha) e como eu executo o terraform init ([dev](https://github.com/chgasparoto/youtube-cleber-gasparoto/blob/main/.github/workflows/0003_tf_apply_dev.yml#L42) e [prod](https://github.com/chgasparoto/youtube-cleber-gasparoto/blob/main/.github/workflows/0003_tf_apply_prod.yml#L39)).

Abraço.

**Comment by fjareta** [comment]
Olá @chgasparoto, estou adorando o curso e aprendendo bastante.
Amigo, tenho uma dúvida a respeito do arquivo main.tf
Após a aula em que criamos nosso primeiro bucket no S3 e que partimos para criar nossa primeira VM, eu tentei criar um outro arquivo chamado vms.tf para não ter que mexer no arquivo main.tf. De inicio ao tentar executar um terraform validate nesse nosso arquivo o programa me apresentou uns erros. Então percebi que esse novo arquivo, vms.tf, não poderia ficar no mesmo diretório que o main.tf.
Movi-o para uma nova pasta acima e tentei executar os comandos e ok, os erros pararam. Mas quando percebi, o terraform ainda não enxerga ele, e os dados na tela apresentados pelo terraform são a respeito do main.tf

Minha pergunta é, como faço para o terraform ler outros arquivos, de eventuais recursos que eu quiser criar sem ter que apagar os dados do main.tf? Fiquei com dúvida nessa parte de arquivos e diretório de como o terraform enxerga e trabalha. Abraços.

**Reply by chgasparoto** [reply]
Olá @fjareta , que bom que está gostando do curso 👍 

Qual erro retornou quando você rodou o `terraform validate`?

> como faço para o terraform ler outros arquivos, de eventuais recursos que eu quiser criar sem ter que apagar os dados do main.tf?

É só criar outros arquivos na mesma pasta que está o seu arquivo `main.tf`. Você não precisa alterar o `main.tf` se não quiser.

> Fiquei com dúvida nessa parte de arquivos e diretório de como o terraform enxerga e trabalha

O Terraform lê todos os arquivos da pasta que está sendo executado e mescla todos eles em um só no runtime, como se fosse apenas um grande arquivo.

Abraço.

**Comment by wesleyosantos91** [comment]
Estou tentando rodar terraform test para validar variáveis e estrutura do meu módulo Terraform, mas ele tenta se conectar à AWS, resultando em erro de credenciais. Existe alguma forma de executar os testes sem precisar de uma conta AWS, apenas validando a estrutura do código e as variáveis?
